### DISCUSSION TITLE
Incorrect Data Testing

### DISCUSSION THREAD

**Discussion by jordanmack** [nervosnetwork/ckb#4734] [Q&A]
I'm working on some testing and I need to create a testnet where there is a node that reports incorrect data to other nodes and also via the RPC. Do we have any tools or existing testing forks that do this?

**Comment by eval-exec** [comment]
Could you clarify what you mean by "incorrect data" in this context?

We don't have a specific tool for this, and you would need to modify the CKB node yourself and compile it to simulate a node that reports incorrect data.

**Comment by doitian** [comment]
Could you provide a brief test case to help us better understand your requirements?

**Comment by gpBlockchain** [comment]
While it's impossible to generate incompatible data directly through RPC, you can employ an incompatible CKB version to create this data beforehand. Afterward, other nodes can be synced with this node to obtain the incompatible data, which will cause the synchronization process to hang. like:https://github.com/cryptape/ckb-py-integration-test/blob/main/test_cases/soft_fork/test_sync_dao_lock_size_mismatch.py

**Comment by jordanmack** [comment]
I want to change specific details that are being reported by the RPC. Particularly, I want to change the CKB amount and the data for specific cells when get_live_cell is called. I want to generate erroneous results for two scenarios:

- A broken or malicious public full node used by dapps.
- A light client that is connected to a broken or malicious node.

Modifying the CKB node with a hard code modification to give incorrect data for a specific cell is an acceptable solution since this is a limited test. A proxy that manipulates data would also be acceptable. I could use suggestions on what the most simple path would be.

**Comment by doitian** [comment]
Another problem is that the node providing invalid data will be banned by other nodes, which leads to a limited window for testing. If you meet such problems, I think we need a config option in CKB to disable banning malicious nodes for testing.

**Reply by jordanmack** [reply]
Say I set up an isolated testnet with a single invalid data node and a single light client that connects to it. Will the light client fully validate a get_live_cell call from a dapp? If so, then only the first scenario I described is needed.

**Reply by doitian** [reply]
@quake @yangby-cryptape Could you help to answer this question?

**Reply by quake** [reply]
The light client will verify the downloaded block via POW, if the node it connects to forges the data, but it does not pass the POW, then the light client ignores the data and disconnects.
